pkg/registry/apis/provisioning/resources/tree.go

```diff
@@ -60,22 +60,38 @@ func (t *FolderTree) DirPath(folder, baseFolder string) (fid Folder, ok bool) {
 	return fid, ok
 }
 
-func FetchRepoFolderTree(ctx context.Context, client *DynamicClient) (*FolderTree, error) {
```

Comment:
Extracted out the usage of the clients here and just us the tree for its tree structure.

Added 3 different constructors to create the tree in different ways and use the tree as convenient data structure. 

-`NewFolderTreeFromUnstructure`: used mainly by exporter as it fetches all folders in the namespace regardless if they are provisioned or not. 
-`NewFolderTreeFromResourceList`: used mainly by the complete sync process because it has to get all the provisioned resources to do reconciliation.  
-`NewEmptyFolderTree`: used mainly by the increment sync to avoid populating the entire tree unnecessarily. The tree will be build as we create folders for the files added or modified.